source/extensions/filters/http/jwt_authn/authenticator.cc

```diff
@@ -0,0 +1,286 @@
+  void sanitizePayloadHeaders(Http::HeaderMap& headers) override;
+
+private:
+  // Fetch a remote public key.
+  void fetchRemoteJwks();
+
+  // Following two functions are for AyncClient::Callbacks
+  void onSuccess(Http::MessagePtr&& response) override;
+  void onFailure(Http::AsyncClient::FailureReason) override;
+
+  // Verify with a specific public key.
+  void verifyKey();
+
+  // Handle the public key fetch done event.
+  void onFetchRemoteJwksDone(const std::string& jwks_str);
+
+  // Calls the callback with status.
+  void doneWithStatus(const Status& status);
+
+  // Return true if it is OK to forward this request without JWT.
+  bool okToBypass();
+
+  // The cluster manager object to make HTTP call.
+  Upstream::ClusterManager& cm_;
+  // The cache object.
+  DataStore& store_;
+
+  // The token data
+  JwtLocationConstPtr token_;
+  // The JWT object.
+  ::google::jwt_verify::Jwt jwt_;
+  // The jwks data object
+  JwksCache::JwksData* jwks_data_{};
+
+  // The HTTP request headers
+  Http::HeaderMap* headers_{};
+  // The on_done function.
+  Authenticator::Callbacks* callback_{};
+
+  // The pending uri_, only used for logging.
+  std::string uri_;
+  // The pending remote request so it can be canceled.
+  Http::AsyncClient::Request* request_{};
+};
+
+void AuthenticatorImpl::sanitizePayloadHeaders(Http::HeaderMap& headers) {
+  for (const auto& rule : store_.config().rules()) {
+    if (!rule.forward_payload_header().empty()) {
+      headers.remove(Http::LowerCaseString(rule.forward_payload_header()));
+    }
+  }
+}
+
+void AuthenticatorImpl::verify(Http::HeaderMap& headers, Authenticator::Callbacks* callback) {
+  headers_ = &headers;
+  callback_ = callback;
+
+  ENVOY_LOG(debug, "Jwt authentication starts");
+  auto tokens = store_.getExtractor().extract(headers);
+  if (tokens.size() == 0) {
+    if (okToBypass()) {
+      doneWithStatus(Status::Ok);
+    } else {
+      doneWithStatus(Status::JwtMissed);
+    }
+    return;
+  }
+
+  // TODO, add supports for multiple tokens.
+  // Only process the first token for now.
+  token_.swap(tokens[0]);
+
+  Status status = jwt_.parseFromString(token_->token());
+  if (status != Status::Ok) {
+    doneWithStatus(status);
+    return;
+  }
+
+  // Check if token is extracted from the location specified by the issuer.
+  if (!token_->isIssuerSpecified(jwt_.iss_)) {
+    ENVOY_LOG(debug, "Jwt for issuer {} is not extracted from the specified locations", jwt_.iss_);
+    doneWithStatus(Status::JwtUnknownIssuer);
+    return;
+  }
+
+  // Check "exp" claim.
+  const auto unix_timestamp = std::chrono::duration_cast<std::chrono::seconds>(
+                                  std::chrono::system_clock::now().time_since_epoch())
+                                  .count();
+  if (jwt_.exp_ < unix_timestamp) {
+    doneWithStatus(Status::JwtExpired);
+    return;
+  }
+
+  // Check the issuer is configured or not.
+  jwks_data_ = store_.getJwksCache().findByIssuer(jwt_.iss_);
```

Comment:
Is the cache only for preconfigured keys or is there an ability to dynamically add to it? Please clarify in comments, thanks.

Reply:
Added comment